### PR TITLE
Config flow for discovered device

### PR DESCRIPTION
**Pull request by filmkorn**
First implementation:
- ✅ Get sensor values
- ✅ Get and set target temperature
- ✅ Start / Stop device (switch)
- ❌ Get ~~/ Set~~ units: You may switch the units on the device for now. Reported temperatures will remain in units as set in hass preferences.
- ✅ Get / Set timer

TODO: 
- Add translation keys to entities.

## Config flow:
image
image
image

## Entities:
image

### CODE REVIEW

**Comment by filmkorn** [comment]
Right now I can't figure out how to get the Coordinator to do anything. The get any data from the device we have to request it - but I don't know where to hook into for this. Currently `custom_components.anova_nano.coordinator.AnovaNanoDataUpdateCoordinator._async_update` is never called.

File: custom_components/anova_nano/translations/en.json [hunk 1]
```diff
@@ -2,17 +2,20 @@
     "config": {
         "step": {
             "user": {
-                "description": "If you need help with the configuration have a look here: https://github.com/ludeeus/anova_nano",
+                "description": "If you need help with the configuration have a look here: https://github.com/mcolyer/hacs_anova_nano",
```
Comment (filmkorn) [inline custom_components/anova_nano/translations/en.json:5]:
```suggestion
                "description": "If you need help with the configuration have a look here: https://github.com/mcolyer/hacs-anova-nano",
```

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
-homeassistant==2024.1.4
-async-timeout==4.0.3
+homeassistant
```
Comment (filmkorn) [inline requirements.txt:1]: Move to pyproject.toml

**Comment by filmkorn** [comment]
Got it to poll the status. Depends on some changes from: https://github.com/filmkorn/pyanova-nano/pull/10

Lots of functionality left:
- Get and set target temperature
- Start / Stop device (switch)
- Get / Set units
- Get / Set timer
Other: 
- Add translation keys.

File: custom_components/anova_nano/sensor.py [hunk 3]
```diff
@@ -1,59 +1,79 @@
...
+        name="internal temperature",
         native_unit_of_measurement=UnitOfTemperature.CELSIUS,
         device_class=SensorDeviceClass.TEMPERATURE,
         state_class=SensorStateClass.MEASUREMENT,
     ),
-    SensorEntityDescription(
+    AnovaSensorEntityDescription(
         key="motor_speed",
+        name="motor speed",
         state_class=SensorStateClass.MEASUREMENT,
```
Comment (filmkorn) [inline custom_components/anova_nano/sensor.py:61]: I wonder what the unit is. When heating the value is around 160 - perhaps rpm?

File: custom_components/anova_nano/sensor.py [hunk 4]
```diff
@@ -1,59 +1,79 @@
...
+        key="internal_temp",
+        name="internal temperature",
         native_unit_of_measurement=UnitOfTemperature.CELSIUS,
         device_class=SensorDeviceClass.TEMPERATURE,
         state_class=SensorStateClass.MEASUREMENT,
     ),
-    SensorEntityDescription(
+    AnovaSensorEntityDescription(
         key="motor_speed",
+        name="motor speed",
```
Comment (filmkorn) [inline custom_components/anova_nano/sensor.py:60]: I wonder what the unit is for the motor speed. When heating the value is around 160 - perhaps rpm?

**Comment by filmkorn** [comment]
@mcolyer Could you give this a try? To install, you copy `custom_components/anova_nano` into `/config/custom_components`.

**Comment by mcolyer** [comment]
Wow this is great, I'll give it a go this weekend.

File: custom_components/anova_nano/manifest.json [hunk 5]
```diff
@@ -3,15 +3,15 @@
...
-  "documentation": "https://github.com/mcolyer/hacs-anova_nano",
+  "documentation": "https://github.com/mcolyer/hacs-anova-nano",
   "iot_class": "local_polling",
-  "issue_tracker": "https://github.com/mcolyer/hacs-anova_nano/issues",
-  "version": "0.0.0",
-  "dependencies": ["bluetooth_adapters"],
-  "requirements": ["pyanova-nano==0.1.0"],
+  "issue_tracker": "https://github.com/mcolyer/hacs-anova-nano/issues",
+  "version": "0.1.0",
+  "requirements": ["git+https://github.com/filmkorn/pyanova-nano.git@dev_testing#pyanova-nano==0.2.0"],
```
Comment (filmkorn) [inline custom_components/anova_nano/manifest.json:10]: Gotta remove the branch once the fixes are better tested and released.

**Comment by mcolyer** [comment]
Okay so I got it to configure my device correctly, however it seems to never get any of the sensor values and then becomes unavailable without me touching anything.

I see this in the core logs:

image

**Comment by filmkorn** [comment]
Thanks for testing! I'm sorry it's not working.
I wonder if it's not using the testing branch [as set in the manifest](https://github.com/filmkorn/hacs-anova-nano/blob/e56ede175d40d0c3e8d574241b0d5f794ad3798b/custom_components/anova_nano/manifest.json#L10). I've had issues with this as well.
Are the logs in Settings > System > Logs showing any additional tracebacks?

**Comment by mcolyer** [comment]
That's the only error that I see.

**Comment by filmkorn** [comment]
Looking into it.

The integration is supposed to be able to be setup without establishing a connection - because I don't want it to remain in an error state until the device is available. This behaviour matches other BLE integrations.

The connection should be established on the first update, but I'm also seeing it fail. Odd because I thought it was working.
In my case it connects on the 2nd or 3rd update but I think the 5 seconds connection timeout might be too short.

In general I find the Nano quite finicky. If you don't disconnect from it explicitly it won't be available for other clients to connect to it. Sometimes I have to turn it on/off by holding the play button or fully restart it even.

**Comment by mcolyer** [comment]
Yep that matches what I'm seeing. The update seems to fail. I tried
unplugging and replugging in the nano and it still doesn't seem to work.

Maybe extend the timeout?

FWIW I'm using an esphome bluetooth to wifi proxy but the setup still seems
to work correctly.

**Comment by filmkorn** [comment]
Rewrote the entire connection logic to use a callback model. Probably not for the last time. But it does works better on my end now.

I do have to resurrect my esphome Bluetooth proxy for more testing.

**Comment by filmkorn** [comment]
Can confirm it's working on my end with an esphome Bluetooth proxy on an a cheap ESP-WROOM-32 board.

If you can find any other error logs I'm keen to try an repro it.

Would you mind giving it another try?

**Comment by mcolyer** [comment]
So I pulled the latest and now get this in the logs:

image

**Comment by mcolyer** [comment]
And it doesn't seem to load

**Comment by filmkorn** [comment]
Can you load the full logs please (either click on the error or find the error in the full log)?

I'm not sure how bluetooth proxies differ from bluetooth adapters. On the esp32 bluetooth proxy I do get this error and I can't find much about it on the internet.
```
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/anova_nano/coordinator.py", line 136, in _async_update_data
await self._connect()
File "/config/custom_components/anova_nano/coordinator.py", line 80, in _connect
await self._client.connect(
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 139, in connect
await self._connect(self._device, timeout_seconds=timeout_seconds)
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 170, in _connect
await self._client.connect()
File "/usr/local/lib/python3.12/site-packages/habluetooth/wrappers.py", line 311, in connect
connected = await super().connect(**kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak/__init__.py", line 615, in connect
return await self._backend.connect(**kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 77, in _async_wrap_bluetooth_operation
return await func(self, *args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 327, in connect
await connected_future
bleak.exc.BleakError: Error ESP_GATT_CONN_FAIL_ESTABLISH while connecting: Connection failed to establish
```

In the logs of the esp32 through https://web.esphome.io/?dashboard_logs I can find:
```
[D][esp32_ble_tracker:270]: Starting scan...
[I][bluetooth_proxy:282]: [0] [58:93:D8:D8:7D:A4] Connecting v3 without cache
[D][esp32_ble_tracker:219]: Pausing scan to make connection...
[I][esp32_ble_client:067]: [0] [58:93:D8:D8:7D:A4] 0x00 Attempting BLE connection
[D][esp-idf:000][BTU_TASK]: W (220139) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[D][esp-idf:000][BTU_TASK]: W (221847) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[D][esp-idf:000][BTU_TASK]: W (222384) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[D][esp-idf:000][BTU_TASK]: W (222594) BT_APPL: gattc_conn_cb: if=3 st=0 id=3 rsn=0x3e

[D][esp-idf:000][BTU_TASK]: W (222596) BT_APPL: gattc_conn_cb: if=4 st=0 id=4 rsn=0x3e

[D][esp-idf:000][BTU_TASK]: W (222596) BT_APPL: gattc_conn_cb: if=5 st=0 id=5 rsn=0x3e

[D][esp-idf:000][BTU_TASK]: W (222597) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e

[D][esp32_ble_client:172]: [0] [58:93:D8:D8:7D:A4] ESP_GATTC_DISCONNECT_EVT, reason 62
[D][esp32_ble_tracker:270]: Starting scan...
```

Reason 62 should be [0x3e: ESP_GATT_CONN_FAIL_ESTABLISH](https://github.com/espressif/esp-idf/blob/46acfdce969f03c02b001fe4d24fa9e98f6adc5e/components/bt/host/bluedroid/api/include/api/esp_gatt_defs.h#L337) - which doesn't help me at all.

I tried to further increase the connection timeout but no success (usually takes about 7 seconds with a Bluetooth adapter according to the logs in with debug logging enabled).

I'm out of ideas honestly. Could be a bug anywhere in pyanova-nano, bleak-esphome, bleak, esphome, esp-idf.

**Comment by filmkorn** [comment]
Somehow it suddenly connected! Got a new error:
```
Traceback (most recent call last):
File "/config/custom_components/anova_nano/coordinator.py", line 140, in _async_update_data
self.status = await self.client.get_sensor_values()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 455, in get_sensor_values
sensor_value_list = await self.send_read_command(ReadCommands.GetSensorValues)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 281, in send_read_command
return await self.send_command(command=command)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 347, in send_command
await get_data()
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line 336, in get_data
await self._client.start_notify(
File "/usr/local/lib/python3.12/site-packages/bleak/__init__.py", line 844, in start_notify
await self._backend.start_notify(characteristic, wrapped_callback, **kwargs)
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 77, in _async_wrap_bluetooth_operation
return await func(self, *args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/client.py", line 646, in start_notify
raise BleakError(
bleak.exc.BleakError: esp32-01-b24244 [B8:D6:1A:B2:42:44]: Nano - 58:93:D8:D8:7D:A4: Notifications are already enabled on service:0e140000-0af1-4582-a242-773e63054c68 characteristic:0e140002-0af1-4582-a242-773e63054c68 handle:14
```

Looks like bleak_esphome complains if you try to subscribe to the same characteristics twice, while backends dont. Will fix that in pyanova_nano.

**Comment by mcolyer** [comment]
Here's the full error:

2024-09-29 08:05:09.641 ERROR (MainThread) [custom_components.anova_nano]
Unexpected error fetching anova_nano data
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py",
line 354, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/anova_nano/coordinator.py", line 136, in
_async_update_data
await self._connect()
File "/config/custom_components/anova_nano/coordinator.py", line 80, in
_connect
await self._client.connect(
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line
139, in connect
await self._connect(self._device, timeout_seconds=timeout_seconds)
File "/usr/local/lib/python3.12/site-packages/pyanova_nano/client.py", line
170, in _connect
await self._client.connect()
File "/usr/local/lib/python3.12/site-packages/habluetooth/wrappers.py",
line 286, in connect
wrapped_backend = self._async_get_best_available_backend_and_device(manager)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/habluetooth/wrappers.py",
line 395, in _async_get_best_available_backend_and_device
raise BleakError(
bleak.exc.BleakError: No backend with an available connection slot that can
reach address C4:64:E3:13:57:9E was found

Let me know when I should try pulling again and trying a fix.

**Comment by filmkorn** [comment]
> bleak.exc.BleakError: No backend with an available connection slot that can reach address C4:64:E3:13:57:9E was found

This likely means that the Bluetooth proxy is not available. I'm having the same issue a lot with my esp32. It's restart button I added shows up as available in hass, but none of my other 3 passive bluetooth devices I have are discovered 90% of the time. They work just fine on a bluetooth adapter.

I've pushed changes that should handle this error more graciously - but it won't solve our issue that the proxy backend is not available.

**Comment by mcolyer** [comment]
I tried updating ESPHome to the latest and moving it to a more central
location, neither seemed to resolve the issue.

Unfortunately my Home Assistant machine doesn't actually have a physical
bluetooth adapter (as it's a virtual machine) so I can't easily test this
via just an adapter.

**Comment by filmkorn** [comment]
I assume your atom Bluetooth proxy work for other devices? 

I think I have to figure out why mine isn't working at all first.
I tried about everything you can find online and nothing has worked. Might have to order a different esp32 before I waste more time on this.

For my own record - will probably create an issue on `esphome`:

1. I don't know which board to use in the config. I have the a "ESP32-CH340C-Type-C" [from aliexpress](https://www.aliexpress.com/item/1005001621773806.html?spm=a2g0o.order_list.order_list_main.5.22661802N2dKmd). Currently using:
     ```yaml
    esphome:
      name: "esp32-01"
      min_version: 2024.6.0  
      name_add_mac_suffix: true    
        
    esp32:
      board: nodemcu-32s
      framework:
        type: esp-idf

1. Scanning: false (according to diagnostics from the ESPHome integration). Looks like the scanner doesn't start automatically even with `continuous: true` in scan_parameters. Had to add:   
     ```yaml
    api:
      encryption:
      on_client_connected:
        - esp32_ble_tracker.start_scan:
           continuous: true
    ```

1. Some api messages are not sent
> [19:50:20][V][api.connection:1477]: Cannot send message because of TCP buffer space 

I can only see this with
```
logger:
  level: VERBOSE
```
But not with:
```
logger:
  level: VERY_VERBOSE
```
Which makes no sense.

This should indicate that the esp32 is out of memory. However I can confirm it is not with a memory template sensor:

```yaml
sensor:
  - platform: template
    name: Free Memory
    lambda: return heap_caps_get_free_size(MALLOC_CAP_INTERNAL);
    unit_of_measurement: 'B'
    state_class: measurement
```

This shows I have around 80-100 kB free - which should be enough.

**Comment by mcolyer** [comment]
So I have one of these
https://shop.m5stack.com/products/atom-lite-esp32-development-kit?ref=NabuCasa

And I have it working with some ble shades that I have and it seemed to
work without issue.

I used the config generated by
https://esphome.io/projects/index.html

I'm not an expert in this area so am not totally confident in my
configuration but it does seem to atleast work for the ble shades.

**Comment by mcolyer** [comment]
I dug a little deeper and saw that esphome uses this configuration:

https://github.com/esphome/bluetooth-proxies/blob/main/m5stack/m5stack-atom-lite.yaml

**Comment by filmkorn** [comment]
Thanks, I'll look into these. I've tried the default esphome BT proxy config and it didn't work either.

The model I have has poor RF performance, can't even connect to the Wifi AP from more than 2 meters away.

Until then Ill put this project on hold. For now - here's some chicken breast being cooked:

**Comment by mcolyer** [comment]
@filmkorn if you release the latest you have I'm happy to merge this in, might as well get it working for some folks even if it doesn't work for me.

**Comment by filmkorn** [comment]
I'd propose to update the readme and add as disclaimer that BT Proxies are not working yet.

On a side note - I flashed a esp32cam with nothing but the BT proxy and that seems to work for passive BT devices.

While the logs for the eps32cam indicate that a connection was established, it doesn't actually make the Nano available.
```
[05:47:14][I][bluetooth_proxy:282]: [0] [58:93:D8:D8:7D:A4] Connecting v3 without cache
[05:47:15][I][esp32_ble_client:067]: [0] [58:93:D8:D8:7D:A4] 0x00 Attempting BLE connection
[05:47:15][I][esp-idf:000][BTU_TASK]: W (21549) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e
[05:47:15][I][esp-idf:000][BTU_TASK]: W (21666) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e
[05:47:16][I][esp-idf:000][BTU_TASK]: W (21759) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x3e
[05:47:17][I][esp32_ble_client:227]: [0] [58:93:D8:D8:7D:A4] Connected
```

This also leads me to to a new error:
```
WARNING (MainThread) [bleak_esphome.backend.client] esp32cam [94:B9:7E:CF:23:B4]: Nano - 58:93:D8:D8:7D:A4: ESPHomeClient bleak client was not properly disconnected before destruction
```

I'll dig into this when I have time but from a quick search, I haven't been able to find much info on this.

**Pull request merged**